### PR TITLE
Initial Conditions may depend on UserObjects

### PR DESCRIPTION
**Pull request by permcody**
Funny, we had the UserObject interface coupled into ICs before, but that doesn't mean that they'd execute before the initial condition. This branch adds the ability to execute UOs _before_ the initial conditions are projected so we can run more complicated algorithms before the ICs are set up. A small test is included but this PR is the precursor to a new system (#8810) that we'll use extensively in the phase-field module.

### CODE REVIEW

File: framework/include/ics/InitialCondition.h [hunk 1]
```diff
@@ -95,6 +97,13 @@ class InitialCondition : public MooseObject,
 
   virtual const std::set<std::string> & getSuppliedItems() override;
 
+  template <typename T>
+  const T & getUserObject(const std::string & name);
```
Comment (aeslaughter) [inline framework/include/ics/InitialCondition.h]: override keyword?

File: framework/include/ics/InitialCondition.h [hunk 2]
```diff
@@ -95,6 +97,13 @@ class InitialCondition : public MooseObject,
...
+  template <typename T>
+  const T & getUserObjectByName(const UserObjectName & name);
```
Comment (aeslaughter) [inline framework/include/ics/InitialCondition.h]: override?

File: framework/include/ics/InitialCondition.h [hunk 3]
```diff
@@ -95,6 +97,13 @@ class InitialCondition : public MooseObject,
...
+
+  const UserObject & getUserObjectBase(const std::string & name);
```
Comment (aeslaughter) [inline framework/include/ics/InitialCondition.h]: override?

**Review by aeslaughter — DISMISSED** [review]
Fine with me.

**Comment by aeslaughter** [comment]
To answer my own question, override doesn't work for template functions, because they can't be virtual.

**Comment by permcody** [comment]
@YaqiWang - This PR may be of interest to you. It adds native capability to MOOSE to allow IC -> UO dependencies. Unfortunately, it totally breaks Rattlesnake right now due to your custom SPH Executioner. Hopefully we can work that out.

**Comment by friedmud** [comment]
This looks good permcody !

**Comment by YaqiWang** [comment]
Yes just looking at the title, I was wondering if I can remove the hack in our SPH executioner. It could be more complicated, I will talk with you when I am back.

**Comment by permcody** [comment]
@YaqiWang or @snschune, How do you want me to handle these changes? This PR properly runs UOs with `execute_on = initial` before any IC that fetches that UO. I looked briefly at the failed test and it looks like the hacked up Executioner is basically doing this manually. I may need some help untangling it. Is it possible to dump that Executioner, or does it do more than just run UOs before ICs? (I haven't really looked at it in depth)

**Comment by YaqiWang** [comment]
Sorry, still trying to pick up things back from the conference. I remember I had to do the hack because of potential cyclic dependency of IC and UO, i.e. the UO depends on a variable, which could be set by a IC and later another IC depends on the UO. I will grab this branch and test it out today.

**Comment by permcody** [comment]
Oh, I may not be able to handle the cyclic dependency if your UO depends on a variable but is also depended upon by another IC. Yuck

**Comment by YaqiWang** [comment]
@permcody the simplest solution is to bypass the dependency setup in the SPHIC by calling `UserObjectInterface::getUserObject` instead of `getUserObject`, which is changed in this PR. Because this is your PR, would you mind to send in a small patch in Rattlesnake?

But your PR has a problem: sometimes ICs requires a user object, whose initial setup is sufficient for the ICs. But you assume here a depending UO must finish its execution for the IC. This cause the failure of some of Rattlesnake SPH tests. IC can call `UserObjectInterface::getUserObject` of cause to bypass this but I think you may want to re-think the default behavior otherwise users could be surprised. Some other SPH test failures are caused by the IC-to-UO dependency (the SPH UO depends on IC but used by setting IC).

**Comment by permcody** [comment]
@YaqiWang - I added one more commit that adds a flag to the IC base class to ignore the UO dependency resolution. I think that this should be a temporary solution. The good news it only requires a single change to Rattlesnake. I have a patch ready.

**Review by dschwen — APPROVED** [review]

**Pull request merged**